File: libgomp/testsuite/libgomp.c-c++-common/target-map-iterators-1c.c
/* { dg-do run } */
/* { dg-require-effective-target offload_device_nonshared_as } */

/* Like target-map-iterators-1.c, but with non-constant iterator
   start/end/step.  */

#include <stdlib.h>

#define DIM1 300
#define DIM2 15

int mkarray (int *x[], int a, int b, int c)
{
  int expected = 0;

  for (int i = a; i < b; i++)
    {
      if ((i + 1) % c == 0)
	{
	  x[i] = (int *) malloc (DIM2 * sizeof (int));
	  for (int j = 0; j < DIM2; j++)
	    {
	      x[i][j] = rand ();
	      expected += x[i][j];
	    }
	}
      else
	x[i] = NULL;
    }

  return expected;
}

int doit (int a, int b, int c)
{
  int *x[DIM1];
  int y;

  int expected = mkarray (x, a, b, c);

  #pragma omp target enter data map(to: x)
  #pragma omp target map(iterator(i=a:b:c), to: x[i][:DIM2])	\
		     map(from: y)
    {
      y = 0;
      for (int i = a; i < b; i++)
	if ((i + 1) % c == 0)
	  for (int j = 0; j < DIM2; j++)
	    y += x[i][j];
    }

  return y - expected;
}

int main (void)
{
  return doit (2, 300, 3);
}

File: libgomp/testsuite/libgomp.fortran/target-map-iterators-1c.f90
! { dg-do run }

! Like target-map-iterators-1.f90, but with non-constant iterator
! start/end/step.

program test
  implicit none

  integer, parameter :: DIM1 = 30
  integer, parameter :: DIM2 = 15

  type :: array_ptr
    integer, pointer :: arr(:)
  end type

  type (array_ptr) :: x(DIM1)
  integer :: expected

  expected = mkarray (2, 29, 3)
  if (doit (2, 29, 3) .ne. expected) stop 1

contains
  integer function mkarray (a, b, c)
    integer :: a, b, c
    integer :: exp = 0
    integer :: i, j

    do i = a, b
      if (mod(i+1, c) == 0) then
	allocate (x(i)%arr(DIM2))
	do j = 1, DIM2
	  x(i)%arr(j) = i * j
	  exp = exp + x(i)%arr(j)
	end do
      end if
    end do

    mkarray = exp
  end function

  integer function doit (a, b, c)
    integer :: a, b, c

    integer:: sum, i, j
    !$omp target map(iterator(i=a:b:c), to: x(i)%arr(:)) map(from: sum)
    sum = 0
    do i = a, b
      if (mod(i+1, c) == 0) then
	do j = 1, DIM2
	  sum = sum + x(i)%arr(j)
	end do
      end if
    end do
    !$omp end target

  doit = sum
  end function
end program
